DataPipelines/mturk_data_pipeline.py

```diff
@@ -105,14 +118,60 @@ def step_2_combine_data(self):
             batch_groups_list.index)
         self.output_data = combined_df
 
-    def step_3_add_timestamp_filters(self, start_time=None, end_time=None):
+    def step_3_add_updated_parameters(self):
+        parametershistory = self.intermediate_data["parameterhistory"]
+        df = self.output_data
+        df = df.reset_index()
+        group_count = 0
+        list_rids_assigned = {}
+        rows = []
+        for p_h in parametershistory.index:
+            history = parametershistory.loc[[p_h]][["parameters"]]
+            update_record = history["parameters"][p_h]["update_record"]
+            for dict_ in update_record:
+                keys = list(dict_.keys())
+                record = df[df["learner_id"] == dict_[keys[0]]]
+                if dict_[keys[0]] not in list_rids_assigned:
+                    list_rids_assigned[dict_[keys[0]]] = []
+                record = record[record[self.intermediate_data["version_json"]] == dict_[keys[1]]]
+                record = record[record["reward"] == dict_[keys[2]]]
+                if record.shape[0] > 1:
+                    record = record.head(1)
+                    if df.index.get_loc(record.iloc[0].name) in list_rids_assigned[dict_[keys[0]]]:
+                        record = record.head(len(list_rids_assigned[dict_[keys[0]]])+ 1).tail(1)
+                if not record.empty:
+                    parameters = history["parameters"][p_h]
+                    parameters = dict(("{}_updated".format(k), v) for k, v in
+                                      parameters.items())
+                    r_i = record["index"].values[0]
+                    list_rids_assigned[dict_[keys[0]]].append(r_i)
+                    row = {}
+                    row["parameters_update"] = parameters
+                    row["update_batch_group"] = group_count
+                    row["index"] = r_i
+                    rows.append(row)
+                else:
+                    row["parameters_update"] = "NA"
```

Comment:
Jiakai pulled the branch and ran the script, and error raised on this line:
UnboundLocalError: local variable 'row' referenced before assignment